# 🐼**Déjame escuchar la música**
---

## 📝**Contenido** <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## **Introducción** <a id='intro'></a>
---


Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.



### 🎯**Objetivo**
Comporbar la siguiente prueba la hipótesis:
- *"La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad"*.


### **Etapas:**
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## **Etapa 1. Descripción de los datos** <a id='data_review'></a>
---


Abre los datos y examínalos.

In [1]:
# Importar pandas
import pandas as pd

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv("music_project_en.csv")

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
display(df.head(10))

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Información general sobre la tabla.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


### **Observaciones:** <a id='data_review_conclusions'></a>
>
>- Contiene siete columnas. 
>- Almacenan los mismos tipos de datos: `object`.
>
>- Según la documentación:
>    - `' userID'`: identificador del usuario o la usuaria;
>    - `'Track'`: título de la canción;
>    - `'artist'`: nombre del artista;
>    - `'genre'`: género de la pista;
>    - `'City'`: ciudad del usuario o la usuaria;
>    - `'time'`: la hora exacta en la que se reprodujo la canción;
>    - `'Day'`: día de la semana.
>
>- Podemos ver tres problemas con el estilo en los encabezados de la tabla:
>    1. Algunos encabezados están en mayúsculas, otros en minúsculas.
>    2. Hay espacios en algunos encabezados.
>    3. Para encabezados con 2 palabras se debe usar "snake_case".
>
>1.   **¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?**
>
>Nuestro DataFrame almacena en su totalidad datos de tipo object, pero llama la atención que la columna `'time'` almacena datos que hacen referencia a un momento exacto de tiempo en relacion a la hora del día. No se trata de un "String" común y corriente.
>
>2.   **¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?**
>
>Se puede inferir que cada registro o fila representa cada ocasión en que se reprodujo una canción en la plataforma de música. Por lo que los datos son suficientes para proporcionar respuestas a nuestra hipótesis.
>
>3.   **¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?**
>
>Gracias al método `info()` se observa que las columnas `'Track'`, `'artist'` y `'genre'` cuentan con datos ausentes por lo que será necesario trabajarlos con el objetivo de que no provoquen un sesgo o afecten nuestro análisis.

[Volver a Contenidos](#back)

## **Etapa 2. Preprocesamiento de datos** <a id='data_preprocessing'></a>
---


El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.

### **Estilo del encabezado** <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

In [6]:
# Bucle en los encabezados 
columns_new = []

for column in df.columns:
    column = (
        column
        .strip() # eliminando los espacios
        .lower() # poniendo todo en minúsculas
    )
    columns_new.append(column)

df.columns = columns_new

# Cambiar el nombre de la columna "userid"
df.rename(columns={'userid':'user_id'}, inplace=True)

print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### **Valores ausentes** <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla.

In [7]:
# Calcular el número de valores ausentes
print(df.isna().sum())


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. La mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [8]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
column_list = ['track', 'artist', 'genre']

for column in df.columns:
    if column in column_list:
        df[column] = df[column].fillna('unknown')

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos.

In [9]:
# Contarprint(df.isna().sum()) valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### **Duplicados** <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [10]:
# Contar duplicados explícitos
print(f'Número de datos duplicados explícitos: {df.duplicated().sum()}')

Número de datos duplicados explícitos: 3826


Elimina todos los duplicados.

In [11]:
# Eliminar duplicados explícitos
df = df.drop_duplicates()

Comprobemos ahora si eliminamos con éxito todos los duplicados.

In [12]:
# Comprobar de nuevo si hay duplicados
print(f'Número de datos duplicados explícitos: {df.duplicated().sum()}')


Número de datos duplicados explícitos: 0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [13]:
# Inspeccionar los nombres de géneros únicos
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Se observan los siguientes duplicados implícitos:

* `hip`
* `hop`
* `hip-hop`

Crear la función `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: lista que contiene todos los valores a reemplazar.
* `correct_genre=`: string para el reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

La función utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [14]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)
    #return df['genre']

In [15]:
# Eliminar duplicados implícitos
replace_wrong_genres(['hip', 'hop', 'hip-hop'], 'hiphop')

Mostrar la lista de valores únicos de la columna `'genre'` una vez más:

In [16]:
# Comprobación de duplicados implícitos
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

>### **Observaciones** <a id='data_preprocessing_conclusions'></a>
>
>- El DataFrame contaba con un número importante de duplicados explícitos, exactamente 3826 lo que representa el 5.87% de la información, por lo que fue muy importante eliminarlos y así evitar que estos datos duplicados interfiriera con el análisis. De esta forma pasamos de contar con 65079 registros a 61253 registros.
>
>- En el caso de los duplicados implícitos es un tema un poco mas complicado. Utilizando la función `unique()` sobre la columna `'genre'` podemos observar que existían variantes de escritura para el mismo género. Por esto fue necesario definir y utilizar una función que estandarizara estas variantes a una sola forma de escribir el genero "hiphop".
>
>- Esta información es muy importante para realizar diferentes tipos de análisis.

[Volver a Contenidos](#back)

## **Etapa 3. Prueba de hipótesis** <a id='hypothesis'></a>
---

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.




El primer paso consiste en evaluar la actividad de los usuarios en cada ciudad. Para ello, se pueden aplicar las etapas dividir-aplicar-combinar vistas previamente: agrupar los datos por ciudad, aplicar el método adecuado para contar y, posteriormente, obtener la cantidad de canciones reproducidas en cada grupo, especificando la columna correspondiente.

El resultado esperado tendría una estructura similar a la siguiente:
`df.groupby(by='....')['column'].method()`



In [17]:
# Contar las canciones reproducidas en cada ciudad
display(df.groupby(by='city', as_index=False)['track'].count())

city  track
0  Shelbyville  18512
1  Springfield  42741

>**Observaciones:**
>
>Se reprodujeron mas canciones en Springfield (42741) que en Shelbyville (18512).
>
>Con el método `groupby()` creamos un grupo para cada ciudad y después con el método `count()` se realizó el conteo de cuantos elementos hay en cadas grupo.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [18]:
# Calcular las canciones reproducidas en cada uno de los tres días
display(df.groupby(by='day',as_index=False)['track'].count())

day  track
0     Friday  21840
1     Monday  21354
2  Wednesday  18059

>**Observaciones:**
>
>El día que mas canciones fueron reproducidas es el Viernes, mientras que el Miércoles se tuvo la menor actividad por parte de los usuarios.

Hasta este punto, ya se ha trabajado con el conteo de entradas agrupadas por ciudad o por día de forma individual. El siguiente paso consiste en desarrollar una función que permita realizar este conteo considerando ambos criterios simultáneamente.

Se propone definir la función `number_tracks()`, cuyo objetivo será calcular el número de canciones reproducidas en un día y ciudad determinados. La función deberá aceptar dos parámetros:

- `day`: día de la semana por el cual se desea filtrar (por ejemplo, `'Monday'`).

- `city`: ciudad a filtrar (por ejemplo, `'Springfield'`).

Dentro de la función, se aplicará un filtrado consecutivo utilizando indexación lógica: primero por día, y posteriormente por ciudad, a partir del subconjunto resultante.

Una vez realizado el filtrado por ambos criterios, se contará el número de valores presentes en la columna `'user_id'`. Este recuento corresponde al número de entradas que se desean obtener. Finalmente, se almacenará el resultado en una variable y se retornará desde la función.

In [19]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_day = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_day_city = df_day[df_day['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    number_tracks = df_day_city['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return number_tracks

Llamando a `number_tracks()` seis veces, cambiando los valores de los parámetros para recuperar los datos de ambas ciudades para cada uno de los tres días.

In [20]:
# El número de canciones reproducidas en Springfield el lunes
print('Número de canciones reproducidas en Springfield el Lunes:', number_tracks('Monday', 'Springfield'))

Número de canciones reproducidas en Springfield el Lunes: 15740


In [21]:
# El número de canciones reproducidas en Shelbyville el lunes
print('Número de canciones reproducidas en Shelbyville el Lunes:', number_tracks('Monday', 'Shelbyville'))

Número de canciones reproducidas en Shelbyville el Lunes: 5614


In [22]:
# El número de canciones reproducidas en Springfield el miércoles
print('Número de canciones reproducidas en Springfield el Miércoles:', number_tracks('Wednesday', 'Springfield'))

Número de canciones reproducidas en Springfield el Miércoles: 11056


In [23]:
# El número de canciones reproducidas en Shelbyville el miércoles
print('Número de canciones reproducidas en Shelbyville el Miércoles:', number_tracks('Wednesday', 'Shelbyville'))

Número de canciones reproducidas en Shelbyville el Miércoles: 7003


In [24]:
# El número de canciones reproducidas en Springfield el viernes
print('Número de canciones reproducidas en Springfield el Viernes:', number_tracks('Friday', 'Springfield'))

Número de canciones reproducidas en Springfield el Viernes: 15945


In [25]:
# El número de canciones reproducidas en Shelbyville el viernes
print('Número de canciones reproducidas en Shelbyville el Viernes:', number_tracks('Friday', 'Shelbyville'))

Número de canciones reproducidas en Shelbyville el Viernes: 5895


>**Conclusiones**
>
>**La hipótesis se acepta parcialmente.**
>
>Los resultados respaldan la declaración pero no de forma absoluta ya que no se ve variación importante en todos los casos, solo en algunos.

[Volver a Contenidos](#back)

# 🧠**Conclusiones** <a id='end'></a>

>Conclusiones sobre la hipótesis:
>
>- En general, los usuarios en la ciudad de Springfield reproducen mas canciones que los usuarios de Shelbyville. Este comportamiento es constante en todos los días de la semana.
>
>- Los días Lunes y Viernes presentan comportamientos similares tanto en Springfield como en Shelbyville.
>
>- Es en el día Miércoles donde se puede observar un comportamiento diferente: Mientras que en Springfield las reproducciones disminuyen, en Shelbyville aumentan.
>
>Los resultados respaldan la declaración pero no de forma absoluta ya que no se ve variación importante en todos los casos, solo en algunos.

### Nota
Este proyecto funje como introducción al manejo de la librería pandas. En la vida real, la prueba de hipótesis estadística es más precisa y cuantitativa.

[Volver a Contenidos](#back)